packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -154,33 +238,277 @@ export class BiconomySmartAccountV2 extends BaseSmartAccount {
+
+      if (this.paymaster && this.paymaster instanceof BiconomyPaymaster) {
+        // Make a call to paymaster.buildTokenApprovalTransaction() with necessary details
+
+        // Review: might request this form of an array of Transaction
+        const approvalRequest: Transaction = await (
+          this.paymaster as IHybridPaymaster<SponsorUserOperationDto>
+        ).buildTokenApprovalTransaction(tokenPaymasterRequest, this.provider)
+        Logger.log('approvalRequest is for erc20 token ', approvalRequest.to)
+
+        if (approvalRequest.data == '0x' || approvalRequest.to == ethers.constants.AddressZero) {
+          return userOp
+        }
+
+        if (!userOp.callData) {
+          throw new Error('Userop callData cannot be undefined')
+        }
+
+        const account = await this._getAccountContract()
+
+        const decodedDataSmartWallet = account.interface.parseTransaction({
+          data: userOp.callData.toString()
+        })
+        if (!decodedDataSmartWallet) {
+          throw new Error('Could not parse call data of smart wallet for userOp')
+        }
+
+        const smartWalletExecFunctionName = decodedDataSmartWallet.name
+
+        if (smartWalletExecFunctionName === 'executeCall') {
+          Logger.log('originally an executeCall for Biconomy Account')
+          const methodArgsSmartWalletExecuteCall = decodedDataSmartWallet.args
+          const toOriginal = methodArgsSmartWalletExecuteCall[0]
+          const valueOriginal = methodArgsSmartWalletExecuteCall[1]
+          const dataOriginal = methodArgsSmartWalletExecuteCall[2]
+
+          batchTo.push(toOriginal)
+          batchValue.push(valueOriginal)
+          batchData.push(dataOriginal)
+        } else if (smartWalletExecFunctionName === 'executeBatchCall') {
+          Logger.log('originally an executeBatchCall for Biconomy Account')
+          const methodArgsSmartWalletExecuteCall = decodedDataSmartWallet.args
+          batchTo = methodArgsSmartWalletExecuteCall[0]
+          batchValue = methodArgsSmartWalletExecuteCall[1]
+          batchData = methodArgsSmartWalletExecuteCall[2]
+        }
+
+        if (approvalRequest.to && approvalRequest.data && approvalRequest.value) {
+          batchTo = [approvalRequest.to, ...batchTo]
+          batchValue = [approvalRequest.value, ...batchValue]
+          batchData = [approvalRequest.data, ...batchData]
+
+          newCallData = await this.encodeExecuteBatch(batchTo, batchValue, batchData)
+        }
+        let finalUserOp: Partial<UserOperation> = {
+          ...userOp,
+          callData: newCallData
+        }
+
+        // Requesting to update gas limits again (especially callGasLimit needs to be re-calculated)
+        try {
+          finalUserOp = await this.estimateUserOpGas(finalUserOp)
+          const cgl = ethers.BigNumber.from(finalUserOp.callGasLimit)
+          if (finalUserOp.callGasLimit && cgl.lt(ethers.BigNumber.from('21000'))) {
+            return {
```

Comment:
Can just return `finalUserOp` here

Reply:
no, don't want finalUserOp with wrong callGasLimit. because it gets modified by estimateUserOpGas, these changes are tested.